iml-system-test-utils/src/lib.rs

```diff
@@ -0,0 +1,56 @@
+pub mod docker;
+pub mod iml;
+pub mod vagrant;
+
+use futures::{future::BoxFuture, FutureExt, TryFutureExt};
+use std::{io, process::ExitStatus, thread, time};
+use tokio::process::Command;
+
+fn handle_status(x: ExitStatus) -> Result<(), io::Error> {
+    if x.success() {
+        Ok(())
+    } else {
+        let err = io::Error::new(
+            io::ErrorKind::Other,
+            format!("process exited with code: {:?}", x.code()),
+        );
+        Err(err)
+    }
+}
+
+pub trait CheckedStatus {
+    fn checked_status(&mut self) -> BoxFuture<Result<(), io::Error>>;
+}
+
+impl CheckedStatus for Command {
+    /// Similar to `status`, but returns `Err` if the exit code is non-zero.
+    fn checked_status(&mut self) -> BoxFuture<Result<(), io::Error>> {
+        println!("Running cmd: {:?}", self);
+
+        self.status()
+            .and_then(|x| async move { handle_status(x) })
+            .boxed()
+    }
+}
+
+pub async fn try_command_n_times(max_tries: u32, cmd: &mut Command) -> Result<(), io::Error> {
+    let mut count = 0;
+    let one_sec = time::Duration::from_millis(1000);
+
+    let mut r = cmd.status().await?;
+
+    // try to run the command max_tries times until it succeeds. There is a delay of 1 second.
+    while !r.success() && count < max_tries {
+        count += 1;
+
+        thread::sleep(one_sec);
+
+        r = cmd.status().await?;
```

Comment:
It could be a "do..while" loop, so you don't need to repeat `cmd.status().await?;`. Ref. https://github.com/rust-lang/rfcs/issues/1313#issuecomment-146665147

Reply:
I think we would still need to call `cmd.status().await?` on each iteration right? If it succeeds on the first call it shouldn't enter the while loop.

Reply:
Yes, but you have the same call before entering the loop, no?

Reply:
I think we can do it this way but i'm not sure how I feel about it. Something feels a little off to me and it might be that I have to declare the mutable variable outside of the loop regardless. This is what I came up with:

```
let mut r: ExitStatus;
while {
    r = cmd.status().await?;
    count += 1;

    thread::sleep(one_sec);

    !r.success() && count < max_tries
} {};

if r.success() {
    Ok(())
} else {
    Err(io::Error::new(io::ErrorKind::Other, format!("Command {:?} failed to succeed after {} attempts.", cmd, max_tries)))
}
```
I'm not seeing much of a gain honestly and it feels a bit hacky to me. Do you still think it's better to do it this way?

Reply:
It's ok. Existing code looks better :)

Reply:
Ok cool, i'll leave it as is then.